WordPress/gutenberg
# Workflows: Add workflow to publish npm packages

## Description
Purports to fix https://github.com/WordPress/gutenberg/discussions/37820. See there for more background.

This PR attempts to automate the publishing of npm packages, as is currently documented in [the docs](https://github.com/WordPress/gutenberg/blob/b54d4d1e20f161309d2c44d7f332a50e9e6b66c4/docs/contributors/code/release.md#packages-releases-to-npm-and-wordpress-core-updates). Prior to this PR, publishing them meant locally running a number of commands on one's system.

As of this PR, this should no longer be necessary (aside from inherently manual processes, e.g. cherry-picking a number of selected commits from one branch to another). Instead, the publishing process can now be started from GitHub.

It is important to keep in mind that there are a [number of different scenarios listed](https://github.com/WordPress/gutenberg/blob/b54d4d1e20f161309d2c44d7f332a50e9e6b66c4/docs/contributors/code/release.md#packages-releases-to-npm-and-wordpress-core-updates) in the docs in which publishing of npms can occur.

For the purpose of this PR, it's especially important to note that some of these scenarios ([Syncing WP trunk](https://github.com/WordPress/gutenberg/blob/b54d4d1e20f161309d2c44d7f332a50e9e6b66c4/docs/contributors/code/release.md#synchronizing-wordpress-trunk), plus arguably [development releases](https://github.com/WordPress/gutenberg/blob/b54d4d1e20f161309d2c44d7f332a50e9e6b66c4/docs/contributors/code/release.md#development-releases)) are supposed to happen basically any time a Gutenberg plugin RC1 is released -- in other words, they should happen _automatically_ (with no user interaction required). For this reason, this PR has them automatically triggered upon release of a Gutenberg plugin RC1.

The other scenarios ([Minor WP releases](https://github.com/WordPress/gutenberg/blob/b54d4d1e20f161309d2c44d7f332a50e9e6b66c4/docs/contributors/code/release.md#minor-wordpress-releases) and [standalone bugfix releases](https://github.com/WordPress/gutenberg/blob/b54d4d1e20f161309d2c44d7f332a50e9e6b66c4/docs/contributors/code/release.md#standalone-bugfix-package-releases)) are of a more manual nature: Someone has to make the decision that a point or bugfix release is required. Consequently, this PR makes it so that these processes can be triggered manually from the corresponding GH Actions page.

## How has this been tested?

We're going to test both the automatically triggered scenario (when an RC1 is published), and the manual one. Both of them share a few common setup steps though.

First of all, apply the following patch on top of this PR's branch in order to disable actual publishing of npms:

```diff
diff --git a/.github/workflows/publish-npm-packages.yml b/.github/workflows/publish-npm-packages.yml
index 7c54dedf30..3aa9d176dc 100644
--- a/.github/workflows/publish-npm-packages.yml
+++ b/.github/workflows/publish-npm-packages.yml
@@ -213,11 +213,7 @@ jobs:
                   minimum_version_bump='patch'
                   if [ '${{ github.event.inputs.release_type }}' = 'next' ]; then
                     sha=$(git rev-parse --short HEAD)
-                    npx lerna version pre"$minimum_version_bump" --preid next."$sha" --no-private
-                    npx lerna publish from-package --dist-tag next
-                  elif [ '${{ github.event.inputs.release_type }}' = 'bugfix' ]; then
-                    npx lerna publish
+                    npx lerna version pre"$minimum_version_bump" --preid next."$sha" --no-private --yes --no-git-tag-version --no-push
                   else
-                    npx lerna version "$minimum_version_bump" --no-private
-                    npx lerna publish from-package
+                    npx lerna version "$minimum_version_bump" --no-private --yes --no-git-tag-version --no-push
                   fi
```

- Make sure you have your own fork of the Gutenberg repo, and that you've added it locally as a remote (E.g. `git remote add ockham git@github.com:ockham/gutenberg.git` if your GH username is `ockham` :slightly_smiling_face: ).
- Check out this PR's branch, and push the contents to your fork's `trunk` branch (you can reset later):
```
git push ockham add/publish-npm-packages-workflow:trunk --force
```
- Make sure that your fork also has the current release branch, plus up-to-date versions of `wp/trunk` and `wp/next`:
```
git checkout release/12.6 && git pull && git push ockham
git checkout wp/next && git pull && git push ockham
git checkout wp/trunk && git pull && git push ockham
```
- Furthermore, you fork also needs to have all tags (for lerna to correctly determine version bumps):
```
git push ockham --tags
```

### Automatic scenario

- Publish an RC1 release for the current release branch in your fork (under 'Releases', e.g. https://github.com/ockham/gutenberg/releases). Specify the correct tag (which will be created by GitHub when publishing the release), e.g. `v12.6.0-rc.1`. Don't forget the leading `v`. Set the release name accordingly (e.g. `12.6.0 RC1`). Make sure to attach a `gutenberg.zip` as an asset (e.g. downloaded from corresponding RC1 release in the upstream Gutenberg repo). Make sure for the upload to finish (progress bar!), tick the 'Prerelease' box, and then publish the release by clicking the green button.
- Go to the 'Actions' tab of your fork's GH page (e.g. https://github.com/ockham/gutenberg/actions).
- In the left sidebar (list of available actions), locate the action whose description reads "Publish npm packages", and click on it.
- You should see the workflow that was triggered by publishing the release at the top of the list. Click on it.
- Inspect the individual steps of the workflow (by expanding them) to verify that they worked.

### Manual scenario

- Go to the 'Actions' tab of your fork's GH page (e.g. https://github.com/ockham/gutenberg/actions).
- In the left sidebar (list of available actions), locate the action whose description reads "Publish npm packages", and click on it.
- Locate the blue "This workflow has a workflow_dispatch event trigger." header, and click the "Run workflow" button next to it.
- Keep the `trunk` branch as the base branch, and enter `next` and `patch` into the text input fields, respectively.

![image](https://user-images.githubusercontent.com/96308/149390667-4070207b-6da1-4d75-be37-853e535af0fe.png)

To be continued

## Screenshots 
TBD

## TODO

- [x] The workflow gets currently stuck at the "Update package versions" step ([see](https://github.com/ockham/gutenberg/runs/4803403085?check_suite_focus=true)). Debug and fix.
- [x] Trigger workflow upon plugin RC release (see https://github.com/WordPress/gutenberg/discussions/37820#discussioncomment-1949319).
- [ ] Correctly set minimum version bump, depending on scenario.
- [x] Cherry-pick commits to `trunk` (see https://github.com/WordPress/gutenberg/discussions/37820#discussioncomment-1949262).
  - [x] Update changelog files
  - [x] Publish
- [ ] Use admin token when syncing to `wp/trunk` and cherry-picking to `trunk`
  - [ ] Make sure to carry over upstream's branch protection rules
- [ ] Use npm token for publishing (see https://github.com/WordPress/gutenberg/discussions/37820#discussioncomment-1961240)
- [ ] When syncing: Apply changes to Core `trunk`
- [ ] Stretch goal: Send Slack notification to #core-editor.

## Types of changes
Automation
